craft_parts/state_manager/reports.py

```diff
@@ -58,3 +63,73 @@ def reason(self) -> str:
             return ""
 
         return "{} changed".format(formatting_utils.humanize_list(reasons, "and", "{}"))
+
+
+class DirtyReport:
+    """The DirtyReport class explains why a given step is dirty.
+
+    A dirty step is defined to be a step that has run, but since doing so one
+    of the following things have happened:
+
+    - One or more properties used by the step have changed.
+    - One of more project options have changed.
+    - One of more of its dependencies have been re-staged.
+    """
+
+    def __init__(
+        self,
+        *,
+        dirty_properties: List[str] = None,
+        dirty_project_options: List[str] = None,
+        changed_dependencies: List[Dependency] = None,
+    ) -> None:
+        """Create a new DirtyReport.
+
+        :param dirty_properties: Properties that have changed.
+        :param dirty_project_options: Project options that have changed.
+        :param changed_dependencies: Dependencies that have changed.
+        """
+        self.dirty_properties = dirty_properties
+        self.dirty_project_options = dirty_project_options
+        self.changed_dependencies = changed_dependencies
+
+    def reason(self) -> str:
+        """Get summarized report.
+
+        :return: Short summary of why the part is dirty.
+        """
+        reasons = []
+
+        reasons_count = 0
+        if self.dirty_properties:
+            reasons_count += 1
+        if self.dirty_project_options:
+            reasons_count += 1
+        if self.changed_dependencies:
+            reasons_count += 1
+
+        if self.dirty_properties:
+            # Be specific only if this is the only reason
+            if reasons_count > 1 or len(self.dirty_properties) > 1:
+                reasons.append("properties")
+            else:
+                reasons.append(f"{self.dirty_properties[0]!r} property")
```

Comment:
It may happen that a lot of properties are different (all properties in the worst case). The `reason()` method is called to populate the "reason" field of an Action:
```
Execute: Skip pull bar (already ran)
Execute: Skip pull foo (already ran)
Execute: Skip build foo (already ran)
Execute: Stage foo (required to build 'bar')
Execute: Build bar
...
```
If this summary is too long, this list will be hard to read. But `OutdatedReport`/`DirtyReport` also has the "complete report" method (not currently implemented), which we can use to have the full list of all changed properties, options and dependencies.

Reply:
Understood, makes sense thanks!